### PR TITLE
feat: Use java.nio.file.Path and support loading of ZIP files to memory

### PR DESCRIPTION
**Pull request by aababilov**
java.nio.file.Path is a more flexible way to retrieve files and
directories.

ZIP file can be loaded to memory instead of being saved to a temporary
directory. This is useful when the ZIP is placed on a non-local file
system or has to be downloaded from the given URL.

### CODE REVIEW

**Comment by aababilov** [comment]
Test loading from URL without a temporary directory:

``` % ./gradlew run  --args="-u https://github.com/MobilityData/gtfs-validator/raw/v1.4.0/usecase/src/test/resources/valid_zip_sample.zip -o $HOME/gtfs-from-url -t 4 --feed_name ca-test-feed"
```

File: core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsZipInMemoryInput.java
Comment (lionel-nj) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsZipInMemoryInput.java:50]: What do you think about renaming this method for clarity as:
```suggestion
    private boolean isInsideZipDirectory(ZipEntry entry) {
```

or as this:
```suggestion
    private boolean isNested(ZipEntry entry) {
```

File: core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsZipInMemoryInput.java
Comment (lionel-nj) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsZipInMemoryInput.java:51]: 👍🏾

File: main/src/main/java/org/mobilitydata/gtfsvalidator/cli/Main.java
Comment (lionel-nj) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/cli/Main.java:82]: Thanks, I did not have the time yet to remove the code duplication #574

**Review by lionel-nj — CHANGES_REQUESTED** [review]
Thanks @aababilov just one comment in line.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/cli/Main.java
Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/cli/Main.java:82]: You are welcome! Now we have a clear case when code duplication should be avoided since there are three different ways to create GtfsInput.

**Comment by aababilov** [comment]
Updated.

**Review by lionel-nj — APPROVED** [review]
Thanks @aababilov, I will wait for @barbeau's final approval to merge this PR.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsInput.java
Comment (barbeau) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsInput.java:43]: Javadoc needs to be updated for abstract class vs. interface

**Review by barbeau — COMMENTED** [review]
@aababilov Nice, LGTM! One comment on documentation. Also, the main README.md and USAGE.md files could use updating to reflect this change too (the README.md doesn't currently even list the `-u` or `-s` parameters).

**Comment by barbeau** [comment]
Actually, looks like there is one [failing test](https://github.com/MobilityData/gtfs-validator/pull/576/checks?check_run_id=1657793723) too after the master branch was merged from another PR:

>org.mobilitydata.gtfsvalidator.cli.CliParametersAnalyzerTest > provideStorageDirectoryCliParameterWithoutSpecifyingUrlShouldReturnFalse FAILED
    org.mockito.exceptions.verification.WantedButNotInvoked at CliParametersAnalyzerTest.java:113

**Comment by barbeau** [comment]
@aababilov I'd also suggest turning off CircleCI builds via the CircleCI console for your fork of this repo. We removed the CircleCI config in this repo and turned off builds here because it was slowing down approvals without any benefits. I'm still getting notifications of failed CircleCI builds on your fork, though, and I believe you can turn these off by disabling CircleCI builds for your fork in the console.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsInput.java
Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsInput.java:43]: An abstract class is still providing an interface, so I would like to keep the existing phasing..

**Comment by aababilov** [comment]
> @aababilov Nice, LGTM! One comment on documentation. Also, the main README.md and USAGE.md files could use updating to reflect this change too (the README.md doesn't currently even list the `-u` or `-s` parameters).

Good point! I would prefer to make it in a separate pull request since options `-u` and `-s` already existed before my commit.

**Comment by aababilov** [comment]
`Test Package Document / test` is failing on `provideStorageDirectoryCliParameterWithoutSpecifyingUrlShouldReturnFalse` which passes locally on my machine. The [test report](https://github.com/MobilityData/gtfs-validator/pull/576/checks?check_run_id=1657793723) suggests me to open file:///home/runner/work/gtfs-validator/gtfs-validator/main/build/reports/tests/test/index.html which is obviously not available.

Can we make the automatic testing platform more easy for debugging?

**Comment by aababilov** [comment]
> @aababilov I'd also suggest turning off CircleCI builds via the CircleCI console for your fork of this repo. We removed the CircleCI config in this repo and turned off builds here because it was slowing down approvals without any benefits. I'm still getting notifications of failed CircleCI builds on your fork, though, and I believe you can turn these off by disabling CircleCI builds for your fork in the console.

Thanks, done!

**Comment by barbeau** [comment]
>Can we make the automatic testing platform more easy for debugging?

@aababilov Agreed that it's not always easy to dig in detail here. I believe @lionel-nj and @fabrice-v set up a packaging of the test results to help with this, although it requires several clicks to download the test report. Here's what I was able to find clicking through some of the commits. @lionel-nj may be able to share more when he's back.

image

**Comment by f8full** [comment]
> a packaging of the test results to help with this,

Yep, see #342 and #455

**Pull request merged**
